### ISSUE TITLE
Contravariant `TypeVar`s don't work in dataclasses

### ISSUE THREAD

**Issue by LeeeeT** [microsoft/pyright#4421]
**Describe the bug**
I'm trying to create a generic dataclass with a contravariant type variable. But I'm getting an error when instantiating the class.

**To Reproduce**
Run pyright for the following code:

```python
from typing import TypeVar, Generic
from dataclasses import dataclass

T_contra = TypeVar("T_contra", contravariant=True)

@dataclass
class MyDataclass(Generic[T_contra]):
    ...

MyDataclass[int]()  # error here
```

```
file.py:10:1 - error: Could not bind method "__new__" because "Type[MyDataclass[int]]" is not assignable to parameter "cls"
    Type "MyDataclass[int]" cannot be assigned to type "MyDataclass[T_contra@MyDataclass]"
      "MyDataclass[int]" is incompatible with "MyDataclass[T_contra@MyDataclass]"
        TypeVar "T_contra@MyDataclass" is contravariant
          "object*" is incompatible with "int" (reportGeneralTypeIssues)
```

**Expected behavior**
No errors.

**Additional context**
pyright: 1.1.287

**Comment by erictraut** [comment]
Thanks for the bug report. This isn't limited to just a dataclass. It occurs with any generic class that defines a `__new__` method with a `cls: type[Self]` parameter.

```python
from typing import Self, TypeVar, Generic

T_contra = TypeVar("T_contra", contravariant=True)

class MyClass(Generic[T_contra]):
    def __new__(cls: type[Self]) -> Self:
        ...

MyClass[int]() # Error
```

This will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.288, which I just published. It will also be included in a future release of pylance.
